### DISCUSSION TITLE
How about merging react-markdown and react-remark to remark-react?

### DISCUSSION THREAD

**Discussion by januswel** [remarkjs/remark#694] [Ideas] [answered]
I found 3 packages rendering Markdown as ReactElement

- remark-react
- react-markdown
- react-remark

It seems maintainers are planning to archive remark-react, but users think this is a canonical package by naming conventions in unified collective

React users needs following APIs

- bear API to deal with Markdown fragments <- react-markdown
    - for dataflow users, like Redux, MobX etc
- custom React Hooks <- react-remark

How about providing the package "remark-react" and merging the essence of react-markdown and react-remark??

**Comment by ChristianMurphy** [comment]
> It seems maintainers are planning to archive remark-react

There are no plans to archive it that I am aware of.

> API to deal with Markdown fragments <- react-markdown

This is where `remark-react` and `rehype-react` fit in, they provide generic JS APIs that can be leveraged, rather than hooks or component specific APIs like `react-markdown` and `react-remark` provide.

> merging the essence of react-markdown and react-remark

Maybe, they're getting closer, especially after `react-markdown` adopted `remark-rehype` to do some of the rendering work.
The remaining different is using `rehype-react` options vs components prop to render custom elements.
It's possible in the next release `react-markdown` could be built on `react-remark`.

It still feels like many of these libraries provide different levels of abstraction.
I could see them becoming better aligned, but am not sure about merging them.

Thoughts? :thought_balloon:

**Reply by wooorm** [reply]
> There are no plans to archive it that I am aware of.

https://github.com/unifiedjs/unified/issues/121

---

The naming of `remark-react` follows the remark plugin naming scheme — that names implies it being a plugin; it shouldn’t be something that isn’t a plugin. I want to deprecate it because it’s better to go remark -> rehype -> react: it allows rehype plugins. The reason it exists is because `rehype` didn‘t exist 6 years ago. And it wasn’t maintained by us then either.

`rehype-react` is a good intermediate plugin. But different from `react-remark` or `react-markdown`.
Also was not maintained by us back when it started. It is or should be used in `react-markdown` and `react-remark`.

`react-markdown` is a recent addition to unified — we’re slowly updating it to better match the rest. However, it has a rather different audience: every new programmer who googles “react” + “markdown” uses it. Folks who don’t know that you can style things with CSS. So, having something for that audience, that’s not too complex, is important.

@ChristianMurphy created `react-remark` as a modern take on what `react-markdown` could be, if I summarise that correctly.

There are a couple of rather weird things in `react-markdown` that can’t easily be added through `react-remark`, but I’m slowly working on cleaning that all up: two majors a year or so. But maybe they’re slightly different, and that’s fine!

**Reply by ChristianMurphy** [reply]
> > There are no plans to archive it that I am aware of.
>
> unifiedjs/unified#121

Ah, I missed that, thanks @wooorm!

> created react-remark as a modern take on what react-markdown could be, if I summarise that correctly.

Exactly, it originally had three goals:
* Leverage `remark-rehype` and `rehype-react` for a more consistent rendering API
* Offer a hooks based API
* Include native TS support

and may expand to include server side rendering support.

> There are a couple of rather weird things in `react-markdown` that can’t easily be added through `react-remark`, but I’m slowly working on cleaning that all up: two majors a year or so. But maybe they’re slightly different, and that’s fine!

agreed :+1:

**Reply by januswel** [reply]
Hmm..., my understating is that,

1. `remark-react` will be deprecated
2. `react-remark` may be merged to `react-markdown`
3. Using `rehype-react` is the canonical way to control some options

Right?

I found the granularities of packages about React may be hard to get for users
Some descriptions that show the current status and future works in the README of each package are very helpful to the users 🤔 


The followings are just my gratitudes 👍🏼 

> `react-markdown`
> every new programmer who googles “react” + “markdown” uses it.
> that’s not too complex, is important.

I got it 👍🏼 
And I will share this with my colleagues that are working with the unified collective

>  I’m slowly working on cleaning that all up

Thank you very much for your works 😆

**Comment by wooorm** [comment, marked as answer]
> I found the granularities of packages about React may be hard to get for users

I think particularly a note in `react-markdown` about when to use `react-remark` instead would be great to have, yeah!

For `remark-react` — the note on switching to `rehype-react` has been there a while already, and I’ll deprecate & archive the project soon; I think it’s okay not to spend too much time on it now!

> Thank you very much for your works 😆

❤️
